### ISSUE TITLE
Bug: Cannot narrow method signature

### ISSUE THREAD

**Issue by aldendaniels** [facebook/flow#3388]
Example:
```javascript
class Animal {
  matches(otherAnimal: Animal) {}
}

class Dog extends Animal {
  matches(otherDog: Dog) {}
}
```

Gives:
```
4:   matches(otherAnimal: Animal) {}
                          ^ Animal. This type is incompatible with the expected param type of
8:   matches(otherDog: Dog) {}
                       ^ Dog```

**Comment by aldendaniels** [comment]
I think this should be allowed, because `Dog` is a subclass of `Animal`, so the parent type constraint is still being met.

**Comment by mwalkerwells** [comment]
``Dog`` is a subclass of ``Animal`` (_a more specific version of Animal_) and it's methods should follow the same pattern: be a more specific version of Animal, or subclass. In order for a function to be a subclass of another, it must be contravariant over its params & covariant over it's return.

````js
const matches1: Animal => void
const matches2: ?Animal => void   // Subclass of matches1
````
https://flowtype.org/docs/variance.html#function-types


**Why?**
What happens in this scenario?
````js
class Animal {
  matches(otherAnimal: Animal) {}
}

class Dog extends Animal {
  breed: string;
  matches(otherDog: Dog) {
    console.log(otherDog.breed)
  }
}

function foo(animal: Animal): void {
    const myAnimal = new Animal()
    // Actually is expecting a dog...
    animal.matches(myAnimal)
}
const myDog = new Dog()
foo(myDog)
````

**Comment by jesseschalken** [comment]
`Dog` could be treated as an `Animal` and thus must be prepared to have an arbitrary `Animal` given into `.matches()`.

What is `matches()` supposed to do? Compare the `Animal` with an arbitrary other `Animal`, or compare an `Animal` with another one of the same species?

If the latter, try this instead:

```js
class Animal<-Species> {
  matches(otherAnimal: Species) {}
}

class Dog extends Animal<Dog> {
  matches(otherDog: Dog) {}
}
```

**Comment by mwalkerwells** [comment]
Can you rephrase & reference lines specifically? It's hard to follow when there's overlap with method names. Do you have any questions regarding why ``foo`` is problematic?

**Comment by mwalkerwells** [comment]
The ``Dog`` method ``matches`` can accept an ``Animal`` or an ``?Animal``. You can call ``this.matches(animal)`` to get the ``Animal`` implementation + add additional logic if needed.

Without more information as to what ``matches`` should do (not sure how you're evaluating a "match" & what you're comparing), it's hard to be more helpful. Definitely not a bug.

**Comment by jesseschalken** [comment]
@mwalkerwells Oh I was just giving an alternative answer to the original question, your answer is fine and correct.

**Comment by vkurchatkin** [comment]
In Flow `extends` implies subtyping, so `Dog.matches` must be a subtype of `Animal.matches`.

**Comment by aldendaniels** [comment]
@jesseschalken @vkurchatkin @mwalkerwells Really great explanations - thank you! This makes complete sense. Sorry for citing this is a bug.

What I really want here, FWIW, is to be able to add more type precision in sub-classes without overriding the parent implementation:

```javascript
class Range {
  contains(range: Range): boolean {
    // Range comparison logic
  }
}

class TextRange extends Range {
  contains(range: TextRange): boolean; // not real syntax
}

class RichTextRange extends Range {
  contains(range: RichTextRange): boolean;  // not real syntax
}

function contains(r1: Range, r2: Range): boolean {
  return r1.contains(r2); // Valid
}

const textRange = new TextRange();
const richTextRange = new RichTextRange();

textRange.matches(richTextRange); // Invalid!
textRange.matches(textRange); // Valid

contains(richTextRange, textRange); // Valid
```

Perhaps the distinction is too subtle, however - there may be better patterns to accomplish the same thing.

**Comment by vkurchatkin** [comment]
Simply remove `contains` from base class

**Comment by aldendaniels** [comment]
> Simply remove contains from base class

The implementation lives on the base class, so I can't remove the method.

I ended up using @jesseschalken suggestion:
```javascript
type RangeType = AbstractRange<*>;

class AbstractRange<Self: RangeType> {
  start: number;
  end: number;

  containsRange(range: Self): boolean {
    ...
  }
}

class TextRange extends AbstractRange<TextRange> {}
class RichTextRange extends AbstractRange<TextRange> {}
```

**Comment by aldendaniels** [comment]
The ability to use `this` in contravariant positions would remove the need for the `Self` generic.
